File: apps/go-sdk/tools.go
package firecrawl

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// AlexandriaExecutionError preserves the request ID for retrying the same payload.
type AlexandriaExecutionError struct {
	RequestID string
	Err       error
}

func (e *AlexandriaExecutionError) Error() string {
	return fmt.Sprintf("%v (request ID: %s)", e.Err, e.RequestID)
}
func (e *AlexandriaExecutionError) Unwrap() error { return e.Err }

func (c *Client) ScrapeAlexandria(ctx context.Context, calls []AlexandriaCall, opts *AlexandriaOptions) (*AlexandriaScrapeData, error) {
	if len(calls) == 0 {
		return nil, &FirecrawlError{Message: "at least one alexandria call is required"}
	}
	if len(calls) > 10 {
		return nil, &FirecrawlError{Message: "at most 10 alexandria calls are allowed per request"}
	}
	for i, call := range calls {
		if strings.TrimSpace(call.Provider) == "" {
			return nil, &FirecrawlError{Message: fmt.Sprintf("alexandria call %d: provider is required", i)}
		}
		if strings.TrimSpace(call.Capability) == "" {
			return nil, &FirecrawlError{Message: fmt.Sprintf("alexandria call %d: capability is required", i)}
		}
	}
	if opts != nil && opts.Timeout != nil {
		if *opts.Timeout <= 0 {
			return nil, &FirecrawlError{Message: "timeout must be positive"}
		}
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(*opts.Timeout+5000)*time.Millisecond)
		defer cancel()
	}

	body := map[string]interface{}{"alexandria": calls}
	mergeOptions(body, opts)

	if _, ok := body["origin"]; !ok {
		body["origin"] = "go-sdk@" + Version
	}
	requestID := ""
	if opts != nil {
		requestID = opts.RequestID
	}
	if requestID == "" {
		var bytes [16]byte
		if _, err := rand.Read(bytes[:]); err != nil {
			return nil, err
		}
		requestID = hex.EncodeToString(bytes[:])
	}
	if !regexp.MustCompile(`^[A-Za-z0-9._:-]{1,128}$`).MatchString(requestID) {
		return nil, &FirecrawlError{Message: "invalid request ID"}
	}
	raw, err := c.http.post(ctx, "/v2/scrape", body, map[string]string{"x-request-id": requestID})
	if err != nil {
		return nil, &AlexandriaExecutionError{RequestID: requestID, Err: err}
	}

	var envelope struct {
		ScrapeID string `json:"scrape_id"`
		Data     *struct {
			Alexandria  []AlexandriaScrapeResult `json:"alexandria"`
			CreditsCost *int                     `json:"creditsCost"`
		} `json:"data"`
	}
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil, &AlexandriaExecutionError{RequestID: requestID, Err: err}
	}
	if envelope.Data == nil || envelope.Data.Alexandria == nil || envelope.Data.CreditsCost == nil || *envelope.Data.CreditsCost < 0 {
		return nil, &AlexandriaExecutionError{RequestID: requestID, Err: &FirecrawlError{Message: "invalid alexandria response"}}
	}
	return &AlexandriaScrapeData{
		RequestID:   requestID,
		ScrapeID:    envelope.ScrapeID,
		Alexandria:  envelope.Data.Alexandria,
		CreditsCost: *envelope.Data.CreditsCost,
	}, nil
}

// FindTools explores the catalogue without executing the tools it returns.
func (c *Client) FindTools(ctx context.Context, opts *FindToolsOptions) (*FindToolsData, error) {
	options := map[string]interface{}{}
	mergeOptions(options, opts)
	result, err := c.ScrapeAlexandria(ctx, []AlexandriaCall{{Provider: "firecrawl", Capability: "find-tools", Options: options}}, nil)
	if err != nil {
		return nil, err
	}
	fail := func(err error) (*FindToolsData, error) {
		return nil, &AlexandriaExecutionError{RequestID: result.RequestID, Err: err}
	}
	if len(result.Alexandria) != 1 {
		return fail(&FirecrawlError{Message: "missing Find Tools result"})
	}
	item := result.Alexandria[0]
	if item.Error != nil {
		status := 0
		if item.Error.Status != nil {
			status = *item.Error.Status
		}
		return fail(&FirecrawlError{StatusCode: status, ErrorCode: item.Error.Code, Message: item.Error.Message})
	}
	raw, err := json.Marshal(item.Data)
	if err != nil {
		return fail(err)
	}
	var data FindToolsData
	if err := json.Unmarshal(raw, &data); err != nil {
		return fail(err)
	}
	return &data, nil
}

File: apps/js-sdk/firecrawl/src/v2/methods/tools.ts
import type {
  AlexandriaCall,
  AlexandriaOptions,
  AlexandriaScrapeData,
  AlexandriaScrapeResult,
  FindToolsData,
  FindToolsOptions,
} from "../types";
import { SdkError } from "../types";
import { HttpClient } from "../utils/httpClient";
import {
  normalizeAxiosError,
  throwForBadResponse,
} from "../utils/errorHandler";

const ALEXANDRIA_MAX_CALLS = 10;

function prepareAlexandriaPayload(
  calls: AlexandriaCall[],
  opts: AlexandriaOptions,
): Record<string, unknown> {
  if (!Array.isArray(calls) || calls.length === 0) {
    throw new Error("alexandria requires at least one call");
  }
  if (calls.length > ALEXANDRIA_MAX_CALLS) {
    throw new Error(`alexandria accepts at most ${ALEXANDRIA_MAX_CALLS} calls`);
  }
  const alexandria = calls.map((call, index) => {
    if (!call || typeof call.provider !== "string" || !call.provider.trim()) {
      throw new Error(`alexandria[${index}].provider cannot be empty`);
    }
    if (typeof call.capability !== "string" || !call.capability.trim()) {
      throw new Error(`alexandria[${index}].capability cannot be empty`);
    }
    if (
      Object.keys(call).some(
        (key) => !["provider", "capability", "options"].includes(key),
      )
    )
      throw new Error("Unknown alexandria call option");
    const item: Record<string, unknown> = {
      provider: call.provider.trim(),
      capability: call.capability.trim(),
    };
    if (call.options != null) {
      if (typeof call.options !== "object" || Array.isArray(call.options)) {
        throw new Error(`alexandria[${index}].options must be an object`);
      }
      item.options = call.options;
    }
    return item;
  });
  if (
    opts.timeout != null &&
    (!Number.isInteger(opts.timeout) || opts.timeout <= 0)
  ) {
    throw new Error("timeout must be a positive integer");
  }
  const payload: Record<string, unknown> = { alexandria };
  if (opts.timeout != null) payload.timeout = opts.timeout;
  if (opts.integration && opts.integration.trim())
    payload.integration = opts.integration.trim();
  if (opts.origin) payload.origin = opts.origin;
  return payload;
}

export async function scrapeAlexandria(
  http: HttpClient,
  calls: AlexandriaCall[],
  opts: AlexandriaOptions = {},
): Promise<AlexandriaScrapeData> {
  const payload = prepareAlexandriaPayload(calls, opts);
  const requestId = opts.requestId ?? crypto.randomUUID();
  if (!/^[A-Za-z0-9._:-]{1,128}$/.test(requestId))
    throw new Error("Invalid requestId");
  try {
    const res = await http.post<{
      success: boolean;
      scrape_id?: string;
      data?: { alexandria?: AlexandriaScrapeResult[]; creditsCost?: number };
      error?: string;
    }>("/v2/scrape", payload, {
      headers: { "x-request-id": requestId },
      // Allow response delivery after the API's capped execution deadline.
      timeoutMs: Math.min(opts.timeout ?? 50000, 50000) + 30000,
    });
    if (res.status !== 200 || !res.data?.success) {
      throwForBadResponse(res, "alexandria");
    }
    const data = res.data.data;
    if (
      !data ||
      !Array.isArray(data.alexandria) ||
      typeof data.creditsCost !== "number" ||
      !Number.isInteger(data.creditsCost) ||
      data.creditsCost < 0
    ) {
      throw new SdkError("Invalid alexandria response");
    }
    return {
      scrapeId: res.data.scrape_id ?? "",
      requestId,
      alexandria: data.alexandria,
      creditsCost: data.creditsCost,
    };
  } catch (err: any) {
    try {
      if (err?.isAxiosError) normalizeAxiosError(err, "alexandria");
      throw err;
    } catch (error) {
      if (error && typeof error === "object")
        Object.assign(error, { requestId });
      throw error;
    }
  }
}

export async function findTools(
  http: HttpClient,
  options: FindToolsOptions = {},
): Promise<FindToolsData> {
  const result = await scrapeAlexandria(http, [
    {
      provider: "firecrawl",
      capability: "find-tools",
      options: { ...options },
    },
  ]);
  const item = result.alexandria[0];
  if (!item) throw new SdkError("Missing Find Tools result");
  if (item.error)
    throw Object.assign(
      new SdkError(item.error.message, item.error.status, item.error.code),
      { requestId: result.requestId },
    );
  return item.data as FindToolsData;
}
